sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -413,29 +415,48 @@ private[spark] class HiveExternalCatalog(client: HiveClient, hadoopConf: Configu
       table
     } else {
       getProviderFromTableProperties(table).map { provider =>
-        assert(provider != "hive", "Hive serde table should not save provider in table properties.")
-        // SPARK-15269: Persisted data source tables always store the location URI as a storage
-        // property named "path" instead of standard Hive `dataLocation`, because Hive only
-        // allows directory paths as location URIs while Spark SQL data source tables also
-        // allows file paths. So the standard Hive `dataLocation` is meaningless for Spark SQL
-        // data source tables.
-        // Spark SQL may also save external data source in Hive compatible format when
-        // possible, so that these tables can be directly accessed by Hive. For these tables,
-        // `dataLocation` is still necessary. Here we also check for input format because only
-        // these Hive compatible tables set this field.
-        val storage = if (table.tableType == EXTERNAL && table.storage.inputFormat.isEmpty) {
-          table.storage.copy(locationUri = None)
+        if (provider == "hive") {
+          val schemaFromTableProps = getSchemaFromTableProperties(table)
+          if (DataType.equalsIgnoreCaseAndNullability(schemaFromTableProps, table.schema)) {
+            table.copy(
+              schema = schemaFromTableProps,
+              provider = Some(provider),
+              partitionColumnNames = getPartitionColumnsFromTableProperties(table),
+              bucketSpec = getBucketSpecFromTableProperties(table),
```

Comment:
yea, it should always be `None`, as we disallow bucketed hive tables in public API. However, we do write bucket spec into table properties, so logically we can support it. I also write a test for it: https://github.com/apache/spark/pull/14750/files#diff-8c4108666a6639034f0ddbfa075bcb37R290